src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp

```diff
@@ -31,6 +34,69 @@ namespace chip {
 namespace app {
 namespace {
 
+/// Handles going through callback-based enumeration of generated/accepted commands
+/// for CommandHandler interface based items.
+///
+/// Offers the ability to focus on some operation for finding a given
+/// command id:
+///   - FindFirst will return the first found element
+///   - FindExact finds the element with the given id
+///   - FindNext finds the element following the given id
+class EnumeratorCommandFinder
+{
+public:
+    enum class Operation
+    {
+        FindFirst, // Find the first value in the list
+        FindExact, // Find the given value
+        FindNext   // Find the value AFTER this value
+    };
+
+    EnumeratorCommandFinder(Operation operation, CommandId target) : mOperation(operation), mTarget(target) {}
+
+    /// Callback to pass in to `Enumerate*` calls in COmmandHandlerInterface
+    CommandHandlerInterface::CommandIdCallback Callback() { return HandlerCallbackFn; }
+
+    /// Get the element found (if any) after the `Callback` is used
+    std::optional<CommandId> GetFound() const { return mFound; }
+
+private:
+    Operation mOperation;
+    CommandId mTarget;
+    std::optional<CommandId> mFound = std::nullopt;
+
+    Loop HandlerCallback(CommandId id)
```

Comment:
So... the existing CommandHandlerInterface API was designed to allow efficiently implementing the one thing it was used for.  Now we are using it for something quite different and the API is a bad match for this, leading to a bunch of extra code and inefficiency.

Followup: we should redesign that API to make it fit the use-case better.

Reply:
I agree - this is inefficient because we want an iterator pattern when a callback pattern exists instead.

Did not try to re-make it here though as that probably requires some extra iterations. At least this is infrequently used and the inefficient path should generally not be used at all (but has to be here for completeness).

Reply:
Created https://github.com/project-chip/connectedhomeip/issues/35790 and added comment about this.

I think currently command iteration is mostly not used, it was added because we may use it for automatic Accepted/Generated commands implementation (and at that point switching the API to be callback based is maybe viable).

Reply:
> At least this is infrequently used

It's used all the time for wildcard subscriptions, no?

> it was added because we may use it for automatic Accepted/Generated commands implementation

Not "may", _do_.  This is how those attributes are implemented, period.